Updated New-ModuleDeployment.ps1

ahmadabdalla opened the pull request:
# Change

Added a custom action to deploy templates across different scopes

## Type of Change

- [x ] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

MrMCake commented:
Small suggestion: 

By fetching the Schema from the ARM template you could automatically determine the scope. One param less to cover ;) As a reference, check line 95 in this [script](https://github.com/MrMCake/Modules/blob/main/.github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1) out. 

Also I'd recommend to move the content of the step into a dedicated script as it makes testing easier and allows you to re-use the code for other purposes.

ahmadabdalla commented:
@MrMCake ready for review

ahmadabdalla commented:
@MrMCake @segraef 

Updated the following: New-ModuleDeployment.ps1 

- Added 'additionalTags' as an optional input. Allows injecting info from the pipeline as tags (i.e. commit id). 
- Support for multiple files as input (comma separated).
- Support for directory-based path (filters for .json files). 
- Support for retry (2 times - 5 seconds apart).
- Moved the 'removeDeployment' logic before the deployment.

Questions:
- Do we need the 'moduleName' parameter? 
- Do we need the 'removeDeployment' parameter?

ahmadabdalla commented on .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1:
```diff
@@ -66,97 +69,123 @@ function New-ModuleDeployment {
-                $parameterFileTags = @{}
+
+        ## Iterate through each file
+        foreach ($file in $parameterFilePath) {
```
Addressed

MrMCake replied:
Approved

ahmadabdalla commented on .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1:
```diff
@@ -66,97 +69,123 @@ function New-ModuleDeployment {
+                    }
+                } #end catch
+            } #end do
+            while ($Stoploop -eq $false) { 
```
Added upper limit as an (or) and added optional parameter, default is 3

MrMCake replied:
Approved

ahmadabdalla commented on .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1:
```diff
@@ -66,97 +69,123 @@ function New-ModuleDeployment {
+        ## Iterate through each file
+        foreach ($file in $parameterFilePath) {
+            $fileProperties = Get-Item -Path $file
+            Write-Verbose "Deploying: $($fileProperties.Name) `n"
```
Resolved

MrMCake replied:
Approved
